common/component/postgresql/awsIAM/aws.go

```diff
@@ -0,0 +1,240 @@
+	err = db.QueryRow(dbExistsCtx, databaseExistsTmpl, dbName).Scan(&dbExists)
+	dbExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to check if the PostgreSQL database %s exists: %v", dbName, err)
+	}
+
+	// Create database if needed using master password in connection string
+	if !dbExists {
+		createDBCtx, createDBCancel := context.WithTimeout(ctx, timeout)
+		_, err := db.Exec(createDBCtx, fmt.Sprintf(createDatabaseTmpl, dbName))
+		createDBCancel()
+		if err != nil {
+			return fmt.Errorf("failed to create PostgreSQL user: %v", err)
+		}
+	}
+	return nil
+}
+
+func CreateUserAndRoleIfNeeded(ctx context.Context, timeout time.Duration, connectionString string, db pginterfaces.PGXPoolConn) error {
+	// Parse database user from connection string
+	dbUser, err := parseUserFromConnectionString(connectionString)
+	if err != nil {
+		return fmt.Errorf("failed to parse database user from connection string to create database %v", err)
+	}
+
+	var userExists bool
+	userExistsCtx, userExistsCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(userExistsCtx, userExistsTmpl, dbUser).Scan(&userExists)
+	userExistsCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		// If user does not exist, create the user with pwd = user name for now.
+		// This will be rotated to use proper iam credentials before connecting again.
+		_, err = db.Exec(ctx, fmt.Sprintf(createUserTmpl, dbUser, dbUser))
+		if err != nil {
+			return fmt.Errorf("failed to create PostgreSQL user: %v", err)
+		}
+	}
+
+	// Check if the role is already granted
+	var roleGranted bool
+	roleGrantedCtx, roleGrantedCancel := context.WithTimeout(ctx, timeout)
+	err = db.QueryRow(roleGrantedCtx, checkRoleTmpl, dbUser).Scan(&roleGranted)
+	roleGrantedCancel()
+	if err != nil && err != pgx.ErrNoRows {
+		return fmt.Errorf("failed to check if the role %v is already granted to the PostgreSQL user %s: %v", awsRole, dbUser, err)
+	}
+
+	// Grant the role if it's not already granted
+	// Note: granting the postgres (ie root) user AWS IAM auth may lock the user out of the database,
+	// so we should prevent this.
+	if !roleGranted && dbUser != "postgres" {
+		grantRoleCtx, grantRoleCancel := context.WithTimeout(ctx, timeout)
+		_, err := db.Exec(grantRoleCtx, fmt.Sprintf(grantRoleTmpl, awsRole, dbUser))
+		grantRoleCancel()
+		if err != nil {
+			return fmt.Errorf("failed to grant PostgreSQL user role: %v", err)
+		}
+	}
+	return nil
+}
+
+func InitAWSDatabase(ctx context.Context, config *pgxpool.Config, timeout time.Duration, connString string, region string, awsAccessKey, awsSecretKey string) error {
+	// Note: check if password supplied, otherwise auto generate the AWS access token immediately
+	var (
+		err       error
+		pass, pwd string
+		db        *pgxpool.Pool
+	)
+	pass, err = parsePasswordFromConnectionString(connString)
+	if err != nil || pass == "" {
+		pwd, err = GetAccessToken(ctx, config.ConnConfig, region, awsAccessKey, awsSecretKey)
```

Comment:
I don't think we need this, because you already have a `BeforeConnect` hook below, so that will set the password